src/coreclr/tools/Common/JitInterface/SwiftPhysicalLowering.cs

```diff
@@ -0,0 +1,182 @@
+                        continue;
+
+                    if (interval.Tag == LoweredType.Float)
+                        loweredTypes.Add(CorInfoType.CORINFO_TYPE_FLOAT);
+
+                    if (interval.Tag == LoweredType.Double)
+                        loweredTypes.Add(CorInfoType.CORINFO_TYPE_DOUBLE);
+
+                    if (interval.Tag == LoweredType.Int64)
+                        loweredTypes.Add(CorInfoType.CORINFO_TYPE_LONG);
+
+                    if (interval.Tag == LoweredType.Opaque)
+                    {
+                        // We need to split the opaque ranges into integer parameters.
+                        // As long as we need to fill more than 4 bytes, we'll split into 8-byte integers.
+                        // If we have more than 2 bytes but less than 4, we'll use a 4-byte integer to represent the rest of the parameters.
+                        // If we have 2 bytes, we'll use a 2-byte integer to represent the rest of the parameters.
+                        // If we have 1 byte, we'll use a 1-byte integer to represent the rest of the parameters.
+                        int intervalOffset = 0;
+                        while (interval.Size - intervalOffset > 4)
+                        {
+                            loweredTypes.Add(CorInfoType.CORINFO_TYPE_LONG);
+                            intervalOffset += 8;
+                        }
+
+                        if (interval.Size - intervalOffset > 2)
+                        {
+                            loweredTypes.Add(CorInfoType.CORINFO_TYPE_INT);
+                            intervalOffset += 4;
+                        }
+
+                        if (interval.Size - intervalOffset > 1)
+                        {
+                            loweredTypes.Add(CorInfoType.CORINFO_TYPE_SHORT);
+                            intervalOffset += 2;
+                        }
+
+                        if (interval.Size - intervalOffset == 1)
+                        {
+                            loweredTypes.Add(CorInfoType.CORINFO_TYPE_BYTE);
+                        }
+                    }
+                }
+                return loweredTypes;
+            }
+        }
+
+        public static CORINFO_SWIFT_LOWERING LowerTypeForSwiftSignature(TypeDesc type)
+        {
+            if (!type.IsValueType || type is DefType { ContainsGCPointers: true })
+            {
+                Debug.Fail("Non-unmanaged types should not be passed directly to a Swift function.");
+                return new() { byReference = true };
+            }
+
+            LoweringVisitor visitor = new(type.Context.Target.PointerSize);
+            visitor.AddFields(type, addTrailingEmptyInterval: false);
+
+            List<CorInfoType> loweredTypes = visitor.GetLoweredTypeSequence();
+
+            // If a type has a primitive sequence with more than 4 elements, Swift passes it by reference.
+            if (loweredTypes.Count > 4)
```

Comment:
Does it make sense to implement an early check in the `GetLoweredTypeSequence` and avoid lowering the full sequence if `loweredTypes.Count > 4`?

Reply:
I think it's fine. The number of frozen structs that would surpass the max size and be passed by-reference is likely very small (as if the type is being passed by-ref, it's not getting most of the benefits that being frozen allows).